# Module 2 Homework - Pandas
Using a dataset from _Wine Spectator_, a wine magazine, we will practice data transformation, grouping, and sorting.

Created by Vibhitha Nandakumar (SJSU ID)  
Last updated: 09/12/2025

In [3]:
csvurl="https://gist.githubusercontent.com/clairehq/" + \
        "79acab35be50eaf1c383948ed3fd1129/raw/407a02139ae1e134992b90b4b2b8c329b3d73a6a/winemag-data-130k-v2.csv"
import pandas as pd
wine = pd.read_csv(csvurl)

**Data cleaning**  
Notice that the first column is redundant. Part of data analysis is cleaning and removing redundancy. How would you drop the redundant column inplace, that is overwrite the dataframe.

In [5]:
wine.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

#### Question 1: ####  
What is the mean of the points column?

In [7]:
wine.describe()

Unnamed: 0        points         price
count  65499.000000  65499.000000  60829.000000
mean   32749.000000     88.434037     35.232932
std    18908.076978      3.030310     39.477858
min        0.000000     80.000000      4.000000
25%    16374.500000     86.000000     17.000000
50%    32749.000000     88.000000     25.000000
75%    49123.500000     91.000000     42.000000
max    65498.000000    100.000000   2500.000000

In [8]:
tot_entries = len(wine)
pts_tot = wine['points'].sum()
pts_mean = pts_tot/tot_entries
print(pts_mean)

88.43403716087269


***The mean is 88.434***

#### Question 2: ####  
How many countries are present in this dataset? (Only count each country once)

In [11]:
wine.iloc[:, 1]

0            Italy
1         Portugal
2               US
3               US
4               US
           ...    
65494       France
65495    Australia
65496           US
65497           US
65498        Spain
Name: country, Length: 65499, dtype: object

In [12]:
unique_countries = wine['country'].unique()
print(len(unique_countries))

42


***42***

#### Question 3: ####
How many times does each country appeared in this dataset? Show each country and the corresponding count (show counts in ascending order)

In [15]:
# print(unique_words)
country_count = {country: (wine['country'] == country).sum() for country in unique_countries}
print(country_count)

{'Italy': 10005, 'Portugal': 2963, 'US': 27177, 'Spain': 3409, 'France': 11174, 'Germany': 1051, 'Argentina': 1907, 'Chile': 2258, 'Australia': 1177, 'Austria': 1635, 'South Africa': 737, 'New Zealand': 733, 'Israel': 259, 'Hungary': 61, 'Greece': 242, 'Romania': 67, 'Mexico': 31, 'Canada': 108, nan: 0, 'Turkey': 43, 'Czech Republic': 6, 'Slovenia': 39, 'Luxembourg': 4, 'Croatia': 44, 'Georgia': 37, 'Uruguay': 61, 'England': 36, 'Lebanon': 20, 'Serbia': 7, 'Brazil': 31, 'Moldova': 30, 'Morocco': 11, 'Peru': 8, 'India': 4, 'Bulgaria': 68, 'Cyprus': 6, 'Armenia': 1, 'Switzerland': 4, 'Bosnia and Herzegovina': 1, 'Ukraine': 5, 'Slovakia': 1, 'Macedonia': 6}


#### Question 4: ####
Create a variable `adjusted_price` containing the adjusted price which is the price subtracted by the average price. *This is called **"centering" transformation** - a method commonly used in the preprocessing step before applying various machine learning algorithms.*

In [17]:
price_tot = wine['price'].sum()
price_mean = price_tot/tot_entries
print(price_mean)

wine['adjusted_price'] = wine.apply(lambda row: row['price']-price_mean, axis=1)

32.72086596741935


In [18]:
wine[['country', 'price', 'adjusted_price']].head()

country  price  adjusted_price
0     Italy    NaN             NaN
1  Portugal   15.0      -17.720866
2        US   14.0      -18.720866
3        US   13.0      -19.720866
4        US   65.0       32.279134

#### Question 5: ####
What is the title of the wine that has the highest points-to-price ratio in the dataset?

In [20]:
wine['pts-to-price'] = wine.apply(lambda row: row['points']/row['price'], axis=1)
wine[['country', 'points', 'price', 'pts-to-price', 'adjusted_price']].head()

country  points  price  pts-to-price  adjusted_price
0     Italy      87    NaN           NaN             NaN
1  Portugal      87   15.0      5.800000      -17.720866
2        US      87   14.0      6.214286      -18.720866
3        US      87   13.0      6.692308      -19.720866
4        US      87   65.0      1.338462       32.279134

In [21]:
max_ratio = wine['pts-to-price'].max()

In [22]:
# max_row = df[df['total_crimes'] == df['total_crimes'].max()]

max_row = wine[wine['pts-to-price'] == max_ratio]
print(max_row)

       Unnamed: 0 country                                        description  \
64590       64590      US  There's a lot going on in this Merlot, which i...   

      designation  points  price    province    region_1          region_2  \
64590         NaN      86    4.0  California  California  California Other   

      taster_name taster_twitter_handle                          title  \
64590         NaN                   NaN  Bandit NV Merlot (California)   

      variety  winery  adjusted_price  pts-to-price  
64590  Merlot  Bandit      -28.720866          21.5  


***Bandit NV Merlot (California)***

#### Question 6: ####
Create a series `flavor_counts` that contains two values: the number of wines that has the word "tart" in the `description` column and the number of wines that has the word "berries" in the `description` column. The index of the Series should be "Tart" and "Berries" for the corresponding values.

In [25]:
# Look for the word "tart", not case sensitive, not a substring of another word.
# Same applies to the word "berries"
# Your code here . . .
words = ['tart', 'berries']

flavor_counts = {word: wine['description'].str.count(word).sum() for word in words}
print(flavor_counts)


{'tart': 4291, 'berries': 3885}


#### Question 7: ####
Let's convert the points into simple star ratings. A score of 90 or higher counts as 3 stars, a score of at least 80 but less than 90 is 2 stars. Any other score is 1 star.

Also, any wines from France should automatically get 3 stars, regardless of points.

Add this new column `star_ratings` to the dataframe with the number of stars for each wine in the dataset.

In [27]:
def rating(x): # x is going to be a value from the Series
    if x >= 90:
        return 3
    elif 80 <= x < 90:
        return 2
    else:
        return 1

scale = wine['points'].map(rating) 
wine['star_ratings'] = scale 
wine.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  adjusted_price  pts-to-price  \
0     White Blend              Nicosia             NaN           NaN   
1  Portuguese Red  Quinta dos Avidagos      -17.720866      5.800000   
2      Pinot Gris            Rainstorm      -18.720866      6.214286   
3        Riesling           St. Julian      -19.720866      6.692308   
4      Pinot Noir         Sweet Cheeks       32.279134      1.338462   

   star_ratings  
0             2  
1             2  
2             2  
3             2  
4             2

#### Question 8: ####
Who are the most common wine reviewers in the dataset? Create a Series whose index is the taster_twitter_handle category from the dataset, and whose values count how many reviews each person wrote.

In [29]:
unique_tasters = wine['taster_twitter_handle'].unique() #creates a list of unique twitter handles

taster_count = {taster: (wine['taster_twitter_handle'] == taster).sum() for taster in unique_tasters}
print(taster_count)
# print(len(taster_count))

{'@kerinokeefe': 5313, '@vossroger': 13045, '@paulgwine\xa0': 4851, nan: 0, '@wineschach': 7752, '@vboone': 4696, '@mattkettmann': 3035, '@wawinereport': 2358, '@gordone_cellars': 2032, '@JoeCz': 2605, '@AnneInVino': 1769, '@laurbuzz': 938, '@worldwineguys': 465, '@suskostrzewa': 593, '@bkfiona': 11, '@winewchristina': 4}


#### Question 9: ####
What combination of countries and varieties are most common? Create a Series whose index is a MultiIndexof {country, variety} pairs. For example, a pinot noir produced in the US should map to {"US", "Pinot Noir"}. Sort the values in the Series in descending order based on wine count.

In [31]:
pair = (wine.groupby(['country', 'variety']).size())
counts = pair.reset_index(name='count')
pair_counts = counts.sort_values('count')
pair_counts_ordered = pair_counts.sort_values(by='count', ascending=False)

print(pair_counts_ordered)

      country                   variety  count
1206       US                Pinot Noir   4918
1099       US        Cabernet Sauvignon   3649
1114       US                Chardonnay   3412
304    France  Bordeaux-style Red Blend   2380
676     Italy                 Red Blend   1870
...       ...                       ...    ...
456   Germany                 Rieslaner      1
711     Italy                 Vespaiolo      1
712     Italy                 Vespolina      1
714     Italy                  Vitovska      1
1004    Spain      Muscat of Alexandria      1

[1304 rows x 3 columns]


#### Question 10 #####
Create a Series whose index is reviewers and whose values is the average score given out by that reviewer. Hint: you will need the `taster_name` and `points` columns.

In [33]:
sums = wine.groupby('taster_name')['points'].sum().to_dict() #geeksforgeeks help for .to_dict()
print(sums)
# print(len(sums))

{'Alexander Peartree': 18063, 'Anna Lee C. Iijima': 188604, 'Anne Krebiehl\xa0MW': 160250, 'Carrie Dykes': 3899, 'Christina Pickard': 358, 'Fiona Adams': 958, 'Jeff Jenssen': 20656, 'Jim Gordon': 180044, 'Joe Czerwinski': 230594, 'Kerin O’Keefe': 471943, 'Lauren Buzzeo': 82386, 'Matt Kettmann': 273214, 'Michael Schachner': 673684, 'Mike DeSimone': 20566, 'Paul Gregutt': 432200, 'Roger Voss': 1156817, 'Sean P. Sullivan': 209075, 'Susan Kostrzewa': 51240, 'Virginie Boone': 419022}
